cpp/job/worker.cpp

```diff
@@ -0,0 +1,122 @@
+// Copyright 2015-2015 the openage authors. See copying.md for legal info.
+
+#include "job_aborted_exception.h"
+#include "job_manager.h"
+#include "thread_id.h"
+#include "worker.h"
+#include "../log.h"
+
+namespace openage {
+namespace job {
+
+Worker::Worker(JobManager *parent_manager)
+		:
+		parent_manager{parent_manager},
+		is_running{false} {
+}
+
+void Worker::start() {
+	this->is_running.store(true);
+	this->executor.reset(new std::thread{&Worker::process, this});
+}
+
+void Worker::stop() {
+	this->is_running.store(false);
+	this->notify();
+}
+
+void Worker::enqueue(std::shared_ptr<JobStateBase> job) {
+	std::unique_lock<std::mutex> lock{this->pending_jobs_mutex};
+	this->pending_jobs.push(job);
+	this->notify();
+}
+
+void Worker::notify() {
+	this->jobs_available.notify_all();
+}
+
+void Worker::join() {
+	this->executor->join();
+}
+
+void Worker::execute_job(std::shared_ptr<JobStateBase> &job) {
+	bool aborted = job->execute([this] () {
+			return not this->is_running.load();
+		});
+	// if the job was not aborted, tell the job manager, that the job has
+	// finished
+	if (not aborted) {
+		this->parent_manager->finish_job(job);
+	}
+}
+
+void Worker::process() {
+	while (this->is_running.load()) {
+		// first fetch jobs from the job manager and execute them
+		auto global_job = this->parent_manager->fetch_job();
+		if (global_job.get() != nullptr) {
+			this->execute_job(global_job);
+			// check after each execution if the worker thread is still running
+			if (not this->is_running.load()) {
+				break;
+			}
+
+			global_job = this->parent_manager->fetch_job();
+		}
+
+
+		// true if jobs from the local job queue should be fetched
+		bool fetch_local = true;
+		// true, if the inner loop should be exited
+		bool escape = false;
+
+		while (fetch_local) {
+			std::unique_lock<std::mutex> lock{this->pending_jobs_mutex};
+			while (this->pending_jobs.empty()) {
+				// wait for new jobs from the local job queue or from the job
+				// manager
+				this->jobs_available.wait(lock);
+				// check if the worker thread is still running after a wake-up
+				if (not this->is_running.load()) {
+					escape = true;
+					break;
+				}
+
+				// check if there is a job from the job manager that has to be
+				// executed
+				global_job = this->parent_manager->fetch_job();
+				if (global_job.get() != nullptr) {
+					// if there is a job, execute it and leave the local
+					// fetching loop, in order to check for more jobs from the
+					// job manager
+					lock.unlock();
+					this->execute_job(global_job);
+					escape = true;
```

Comment:
`escape = false;`
also, this is redundant with the previous query. should be avoided by `first` execution check which bypasses the `wait`